### ISSUE TITLE
IDE-friendly registration of links, parameters, and persistent values

### ISSUE THREAD

**Issue by beam2d** [pfnet/chainer#1970]
This is a proposal of new interface of registering links, parameters, and persistent values to links and chains. The current interface (add_param, add_persistent, and add_link) automatically adds an attribute to the link object, which is harmful for writing codes with IDE like PyCharm, since the attribute is not recognized by them. I want to support these objects explicitly substituted to the attributes so that PyCharm can recognize them, and then we can utilize the code-completion and type checking (though the latter needs type hinting annotations to the Chainer's internal code).

The detailed design is described here: https://docs.google.com/document/d/1SS-N8kT-2tZpKnrKBKeVf4tSE_QM9YhsbC4SRe4Klf0/edit?usp=sharing. I'm wondering whether we should deprecate the existing methods or not. I'm planning to make this change for v2, though the compatibility-breaking change is small and it might be ok to include other parts to v1.

**Comment by jekbradbury** [comment]
There's another option: just override `Chain.__setattr__` so it checks if the object being added is an instance of `Link`; if it is, add it as a sub-link; if it isn't, add it as an ordinary Python attribute. This way it would be possible to write `__init__` functions like this:

```python
def __init__(self, size):
    super().__init__()
    self.l1 = L.Linear(size, size)
    self.l2 = L.Linear(size, 2)
```

**Comment by beam2d** [comment]
That looks much more natural than one I gave. One concern I have is whether there is a use case that one wants to add a link/variable object as an attribute without putting it into the link/parameter hierarchy. A workaround exists for this case: wrapping the link/variable object by another object which is not an instance of Link/Variable.

If the above option is acceptable, is it enough to just iterate over `__dict__` when one wants to enumerate all child links or all parameters? I'm thinking about reorganizing the optimization stuffs (in parallel to this issue), and if we can avoid enumerating all parameters in each iteration, this option might be acceptable, too.

**Comment by jekbradbury** [comment]
That sounds like it should work, although another option is as follows:

Keep using a `_children` attribute for iteration, and add to that during `__setattr__`, but don’t add to it if the Link is added using a `add_without_registration` method.

**Comment by beam2d** [comment]
I'm thinking about a slight modification to my original proposal that removes the redundancy of writing the name twice. The link/parameter is first registered without names (e.g. the name is set to the empty string by default). The `__setattr__` method checks if the argument has been registered without names, and in that case it changes the name to the attribute name.

**Comment by delta2323** [comment]
Fixed via #2657.
